src/Measurements.jl

```diff
@@ -0,0 +1,377 @@
+# following functions.
+
+abstract type AbstractMeasurement end
+
+
+"""
+    measure!(measurement, mc, model, sweep_index)
+
+Performs a measurement during the measurement phase.
+"""
+function measure!(m::AbstractMeasurement, mc, model, i)
+    throw(MethodError(measure!, (m, mc, model)))
+end
+
+
+###################################
+# You may also implement
+
+
+"""
+    prepare!(measurement, mc, model)
+
+This method is called between the thermalization and the measurement phase.
+"""
+prepare!(m::AbstractMeasurement, mc, model) = nothing
+
+
+"""
+    finish!(measurement, mc, model)
+
+Finish a measurement. This method is called after the measurement phase.
+"""
+finish!(m::AbstractMeasurement, mc, model) = nothing
+
+
+"""
+    observables(measurement)
+
+Returns a dictionary of observables.
+
+The default implementation searches for all fields `<: AbstractObservable` and
+builds a dictionary pairs `name(obs) => obs`.
+"""
+function observables(m::AbstractMeasurement)
+    fns = fieldnames(typeof(m))
+    os = [getfield(m, fn) for fn in fns if getfield(m, fn) isa AbstractObservable]
+    Dict{String, AbstractObservable}(MonteCarloObservable.name(o) => o for o in os)
+end
+
+
+"""
+    save!(measurement, filename, entryname)
+
+Saves a measurement to a jld-file `filename` in group `entryname`.
+
+The default implementation uses `observables(measurement)` to find observables
+and saves them using
+`MonteCarloObservable.saveobs(obs, filename, entryname * "/" * obs.group)`.
+The `entryname` passed to this function is equivalent to the structure given by
+`measurements()`, e.g. `ME/config/` if `measurements(mc)[:ME][:config]` carries
+a measurement.
+
+See also [`save_measurements!`](@ref), [`observables`](@ref)
+"""
+function save!(m::AbstractMeasurement, filename::String, entryname::String)
+    for (k, obs) in observables(m)
+        saveobs(obs, filename, entryname * "/" * obs.group)
+    end
+    nothing
+end
+
+
+################################################################################
+# A new model may implement the following for convenience
+
+
+"""
+    default_measurements(mc, model)
+
+Return a dictionary of default measurements for a given Monte Carlo flavour and
+model. If there is no implementation given for the specific Monte Carlo flavour
+an empty dictionary will be returned.
+"""
+default_measurements(mc, model) = Dict{Symbol, AbstractMeasurement}()
+
+
+################################################################################
+# mc based, default measurements
+
+
+"""
+    ConfigurationMeasurement(mc, model[, rate=1])
+
+Measures configurations of the given Monte Carlo flavour and model. The rate of
+measurements can be reduced with `rate`. (e.g. `rate=10` means 1 measurement per
+10 sweeps)
+"""
+struct ConfigurationMeasurement{OT <: Observable} <: AbstractMeasurement
+    obs::OT
+    rate::Int64
+    ConfigurationMeasurement(mc, model, rate=1) = new(
```

Comment:
Create `o = Observable(...)` first and then use `new{typeof(o)}(...)` to fix the issue.